Indicators/Covariance.cs

```diff
@@ -0,0 +1,197 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using QuantConnect.Data.Market;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// TODO: figure out what to write here
+    /// </summary>
+    public class Covariance : BarIndicator, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// RollingWindow to store the data points of the target symbol
+        /// </summary>
+        private readonly RollingWindow<decimal> _targetDataPoints;
+
+        /// <summary>
+        /// RollingWindow to store the data points of the reference symbol
+        /// </summary>
+        private readonly RollingWindow<decimal> _referenceDataPoints;
+
+        /// <summary>
+        /// Symbol of the reference used
+        /// </summary>
+        private readonly Symbol _referenceSymbol;
+
+        /// <summary>
+        /// Symbol of the target used
+        /// </summary>
+        private readonly Symbol _targetSymbol;
+
+        /// <summary>
+        /// Covariance of the target used in relation with the reference
+        /// </summary>
+        private decimal _covariance;
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; private set; }
+
+        /// <summary>
+        /// Gets a flag indicating when the indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady => _targetDataPoints.Samples >= WarmUpPeriod && _referenceDataPoints.Samples >= WarmUpPeriod;
+
+
+        /// <summary>
+        /// Creates a new Covariance indicator with the specified name, target, reference,  
+        /// and period values
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="targetSymbol">The target symbol of this indicator</param>
+        /// <param name="period">The period of this indicator</param>
+        /// <param name="referenceSymbol">The reference symbol of this indicator</param>
+        public Covariance(string name, Symbol targetSymbol, Symbol referenceSymbol, int period)
+            : base(name)
+        {
+            // Assert the period is greater than two, otherwise the covariance can not be computed
+            if (period < 2)
+            {
+                throw new ArgumentException($"Period parameter for Covariance indicator must be greater than 2 but was {period}");
+            }
+
+            WarmUpPeriod = period + 1;
```

Comment:
Mind explaining why it requires 1 extra value than the requested period?

Reply:
Sorry, I was taking the lead from another indicator without considering if it's relevant here. I'll remove the plus 1